src/main/java/org/opentripplanner/middleware/bugsnag/jobs/BugsnagEventJob.java

```diff
@@ -0,0 +1,112 @@
+import static org.opentripplanner.middleware.spark.Main.getConfigPropertyAsInt;
+
+/**
+ * Job to check for new event data requests and confirm that Bugsnag has completed the request. At which point obtain
+ * the event data from Bugsnag storage, save to Mongo and remove stale events.
+ */
+public class BugsnagEventJob implements Runnable {
+
+    private static final int BUGSNAG_REPORTING_WINDOW_IN_DAYS
+        = getConfigPropertyAsInt("BUGSNAG_REPORTING_WINDOW_IN_DAYS", 14);
+
+    private static TypedPersistence<BugsnagEventRequest> bugsnagEventRequests = Persistence.bugsnagEventRequests;
+    private static TypedPersistence<BugsnagEvent> bugsnagEvents = Persistence.bugsnagEvents;
+
+    /**
+     * On each cycle get the newest event request. Update the events based on the response from Bugsnag and remove
+     * stale events.
+     */
+    public void run() {
+        Bson filter = Filters.ne("status", "complete");
+        BugsnagEventRequest originalRequest =
+            bugsnagEventRequests.getOneFiltered(filter, Sorts.descending("dateCreated"));
+        if (originalRequest != null) {
+            manageEvents(originalRequest);
+        }
+        removeStaleEvents();
+    }
+
+    /**
+     * Confirm that the event request has completed and update event data accordingly
+     */
+    private void manageEvents(BugsnagEventRequest originalRequest) {
+
+        BugsnagEventRequest currentRequest = BugsnagDispatcher.getEventDataRequest(originalRequest);
+        if (!currentRequest.status.equalsIgnoreCase("completed")) {
+            // request not completed by Bugsnag yet
+            return;
+        }
+
+        removeStaleEventRequests(originalRequest);
+
+        // get event data produced from original event request from Bugsnag storage
+        List<BugsnagEvent> events = BugsnagDispatcher.getEventData(currentRequest);
+
+        // add new events
+        for (BugsnagEvent bugsnagEvent : events) {
+            Set<Bson> clauses = new HashSet<>();
+            clauses.add(eq("errorId", bugsnagEvent.errorId));
+            clauses.add(eq("projectId", bugsnagEvent.projectId));
+            clauses.add(eq("receivedAt", bugsnagEvent.receivedAt));
+            if (bugsnagEvents.getOneFiltered(Filters.and(clauses)) == null) {
+                // only create event if not present
+                bugsnagEvents.create(bugsnagEvent);
+            }
+        }
+    }
+
+    /**
+     * Remove event requests which have been completed or superseded
+     */
+    private void removeStaleEventRequests(BugsnagEventRequest latestRequest) {
+        Bson filter = Filters.lte("dateCreated", latestRequest.dateCreated);
+        List<BugsnagEventRequest> eventRequests = bugsnagEventRequests.getFiltered(filter);
+        for (BugsnagEventRequest eventRequest : eventRequests) {
+            // remove event request now that it has been completed
+            bugsnagEventRequests.removeById(eventRequest.id);
+        }
```

Comment:
Is there a way to bulk delete here instead of finding and then deleting one-by-one?

Reply:
Updated.